##  Pandas Data Transformation 
#### Collapse multiple columns into a single column

This code reads a CSV file and collapses all the columns 

In [9]:
import pandas as pd
import numpy as np
df = pd.read_csv('inspections.csv', na_filter=False,low_memory=False)
df.head()

Inspection ID Violation ID1 Violation ID2 Violation ID3 Violation ID4  \
0       1970968                                                           
1       1970956            18            18                               
2       1970947                                                           
3       1970940                                                           
4       1970935            34            34            33                 

  Violation ID5 Violation ID6 Violation ID7 Violation ID8 Violation ID9  \
0                                                                         
1                                                                         
2                                                                         
3                                                                         
4                                                                         

  Violation ID10 Violation ID11 Violation ID11.1 Violation ID12  \
0                                                                 
1                                                                 
2                                                                 
3                                                                 
4                                                                 

  Violation ID13 Violation ID14 Violation ID15 Violation ID16 Violation ID17  \
0                                                                              
1                                                                              
2                                                                              
3                                                                              
4                                                                              

  Violation ID18  
0                 
1                 
2                 
3                 
4

We want to collapse all the ViolationID values columns into a single 'Count' column as follows

<table>
    <th>InspectionID</th><th>Count</th><th>ViolationID</th>
    <tr><td>1970956</td><td>18</td><td>ViolationID1</td></tr> 
    <tr><td>1970935</td><td>34</td><td>ViolationID1</td></tr> 
    <tr><td>1970933</td><td>37</td><td>ViolationID1</td></tr> 
    <tr><td>1970930</td><td>38</td><td>ViolationID1</td></tr> 
    <tr><td>1970925</td><td>34</td><td>ViolationID1</td></tr> 
</table>

In [10]:
cols=['Inspection ID', 'Count', 'ViolationID']
df1 = pd.DataFrame(columns=cols)
it = iter(df.columns)
next(it, None)  # skip first column
for col in it:
    df2 = df[['Inspection ID',col]]
    m = [col]*len(df2.index)
    df2["ViolationID"] = m
    df2.columns = cols
    #print (df2.head())
    df1 = df1.append(df2, ignore_index=True)
df1 = df1[df1['Count'] != '']
#df1 = df1.sort_values(by='Inspection ID') 
print (df1.head())
print ("Generating output CSV ..")
df1.to_csv("transformed_inspections.csv", sep=',')
print ("File generated")

C:\Anaconda3\lib\site-packages\ipykernel\__main__.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy


  Inspection ID Count    ViolationID
1       1970956    18  Violation ID1
4       1970935    34  Violation ID1
5       1970933    37  Violation ID1
6       1970930    38  Violation ID1
8       1970925    34  Violation ID1
Generating output CSV ..
File generated


In [4]:
#Interating over all rows (as below) is slow and should be avoided
'''
i = 0
df1 = pd.DataFrame(columns=['InspectionID', 'ViolationID', 'Count'])
for index, row in df.iterrows():
    it = iter(df.columns)
    next(it, None)  # skip first item.
    for col in it:
        if (row[col] != ''):
            i = i + 1
            df1.loc[i] = [str(row['Inspection ID']), col, row[col]]
        else:
            break
df1.head()
df1.to_csv("transformed_inspections.csv", sep=',')
'''

'\ni = 0\ndf1 = pd.DataFrame(columns=[\'InspectionID\', \'ViolationID\', \'Count\'])\nfor index, row in df.iterrows():\n    it = iter(df.columns)\n    next(it, None)  # skip first item.\n    for col in it:\n        if (row[col] != \'\'):\n            i = i + 1\n            df1.loc[i] = [str(row[\'Inspection ID\']), col, row[col]]\n        else:\n            break\ndf1.head()\ndf1.to_csv("transformed_inspections.csv", sep=\',\')\n'